### PR TITLE
Issue #5048 - Review temporary buffer usage

### LINKED ISSUE #5048: Review temporary buffer usage
**Jetty version**
9.4.x, 10.0.x

**Description**

See #5045 for some background.   After looking at that issue, it is apparent that when we fail to do all our writing as channels, we are probably doing a very bad job in regards to too many copies into temporary buffers and not reusing those temporary buffers:

- The [HttpContentRangeWriter](https://github.com/eclipse/jetty.project/blob/fb4cfa9be6c45d36bb85809f96de5433087b09e3/jetty-server/src/main/java/org/eclipse/jetty/server/resource/HttpContentRangeWriter.java#L49) is preferring a direct buffer, but [ByteBufferRangeWriter](https://github.com/eclipse/jetty.project/blob/fb4cfa9be6c45d36bb85809f96de5433087b09e3/jetty-server/src/main/java/org/eclipse/jetty/server/resource/ByteBufferRangeWriter.java#L62) always does a writeTo, causing a copy to a temp buffer.    Even if an indirect buffer is provided, it is likely to be a readonly buffer and same problem.
- [CachedHttpContent](https://github.com/eclipse/jetty.project/blob/jetty-9.4.30.v20200611/jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java#L585) always returns a readOnlyBuffer for in memory content. This prevents access to the underlying array and thus can cause more copying.   We should consider the risks of returning a normal buffer... or making our own readonly wrapper that still exposes the array.
- SeekableByteChannelRangeWriter needs to be reviewed.  If we can only write ranges as byte[] then reading them as ByteBuffers will only require yet another copy to the byte array.

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Implementations of `HttpContent` now use the servers `ByteBufferPool` to allocate temporary buffers for resources.
Any allocated buffers are released back to the `ByteBufferPool` when the `HttpContentrelease()` method is called.

Avoid using direct buffers before calling `BufferUtil.writeTo()`.

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/CompressedContentFormat.java [hunk 1]
```diff
@@ -50,7 +50,12 @@ public boolean equals(Object o)
         if (_extension == null && ccf._extension != null)
             return false;
 
-        return _encoding.equalsIgnoreCase(ccf._encoding) && _extension.equalsIgnoreCase(ccf._extension);
+        return equalsIgnoreCase(_encoding, ccf._encoding) && equalsIgnoreCase(_extension, ccf._extension);
+    }
+
+    private static boolean equalsIgnoreCase(String a, String b)
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/CompressedContentFormat.java]: Isn't there a `StringUtil.equalsIgnoreCase` method?  If not, this method should go there.

File: jetty-http/src/main/java/org/eclipse/jetty/http/ResourceHttpContent.java [hunk 2]
```diff
@@ -156,7 +151,26 @@ public ByteBuffer getIndirectBuffer()
             return null;
         try
         {
-            return BufferUtil.toBuffer(_resource, false);
+            ByteBuffer byteBuffer = ByteBufferPoolUtil.resourceToBuffer(_resource, false, _bufferPool);
+            _allocatedBuffers.add(byteBuffer);
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/ResourceHttpContent.java:155]: What is the lifecycle of a `ResourceHttpContent`? Can it be put into a cache?  If so, could `getIndirectBuffer` be called again and again, and the list of allocated buffers get longer and longer with release never called or called in the distant future?    Is release always called?

I'd be perhaps more included to have a `ByteBuffer getIndirectBuffer(ByteBufferPool pool)` method that the caller can pass in their pool and then be responsible for doing the release themselves.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPoolUtil.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            buffer = direct ? BufferUtil.allocateDirect(ilen) : BufferUtil.allocate(ilen);
+
+        int pos = BufferUtil.flipToFill(buffer);
+        if (resource.getFile() != null)
+            BufferUtil.readFrom(resource.getFile(), buffer);
+        else
+        {
+            try (InputStream is = resource.getInputStream())
+            {
+                BufferUtil.readFrom(is, ilen, buffer);
```
Comment (gregw) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPoolUtil.java:56]: It is the temporary buffers like the one used inside `readFrom` that this issue was initially raised about.

File: jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java [hunk 4]
```diff
@@ -674,10 +674,10 @@ protected boolean sendData(HttpServletRequest request,
                 content.getResource().writeTo(out, 0, content_length);
             }
             // else if we can't do a bypass write because of wrapping
-            else if (written || !(out instanceof HttpOutput))
+            else if (written)
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:677]: why was the instanceof test removed?  The response object may be wrapped!

File: jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPoolUtil.java (hunk 3 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPoolUtil.java:56]: Also I think we can use our context to use bigger temp buffers inside `readFrom`

**Review by gregw — CHANGES_REQUESTED** [review]
Not exactly what I was thinking.   Let's hangout.

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java [hunk 5]
```diff
@@ -237,7 +255,7 @@ private HttpContent load(String pathInContext, Resource resource, int maxBufferS
                 {
                     String compressedPathInContext = pathInContext + format._extension;
                     CachedHttpContent compressedContent = _cache.get(compressedPathInContext);
-                    if (compressedContent == null || compressedContent.isValid())
+                    if (compressedContent == null || !compressedContent.isValid())
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java:258]: Why was this changed?

Comment (lachlan-roberts) [reply jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java:258]: We only want to get the resource from the factory again if it is null or not valid.
The code doesn't make much sense the way it was, I think it was a mistake and someone forgot the `!`.
